Upstream blink16 TUI enhancements & add BIOS video ram support

ghaerr opened the pull request:
This PR sends various `blink16` enhancements upstream to `blink`.

Fixes TUI behavior when mixing c)ontinue and C)ontinue harder TUI commands.

Adds D)isplay only command for fast display of pty/teletypewriter output only. (Just like C but only pty panel, ^C to stop. Very useful for much faster boot of operating systems before starting a debug session).

Draws pty output and accepts pty input during C)ontinue and D)isplay commands.

Adds `int 0x16` AH=1 check keypress BIOS function.

There remains a bit of funkiness having to do with the `setitimer` real time timer and ALARM status line redraw callbacks during C) and D) execution, still perfecting that.

With these changes, `blink` can now boot and run both the [ELKS](https://github.com/jbruchon/elks) kernel and [FreeDOS](https://www.freedos.org/download/) v1.3 in real mode. (These are the `fd1440.img` and `freedos13.img` over at [blink16](https://github.com/ghaerr/blink16)).

Here's `blink` booting ELKS: 
![blink booting ELKS](https://user-images.githubusercontent.com/11985637/234748139-7f3a89dd-5375-4cec-b697-04f4861254cb.png)
(Note: the ELKS bioshd retries and errors are the result of `blink` not yet implementing `int 0x13` AH=3 Write Sectors.

Here's the same ELKS boot running in D)isplay only mode:
![blink D command](https://user-images.githubusercontent.com/11985637/234748830-892a8b0b-8615-4633-be8f-255f7959f998.png)

tkchia commented:
Hello @ghaerr,

Incidentally, after seeing the CP437 dumps of the interrupt vector table, I now recall that ELKS reads and copies out the [diskette parameter table](https://www.delorie.com/djgpp/doc/rbinter/id/55/24.html) (the `int 0x1e` vector), so that it can modify it as needed.  I am now thinking if I should try to add a simple parameter table to Blink's default "BIOS" (`defbios.c`).

Thank you!

ghaerr commented:
Hello @tkchia,

> I am now thinking if I should try to add a simple parameter table to Blink's default "BIOS" (defbios.c).

Yes, that is definitely needed. Currently ELKS is copying garbage from the uninitialized (likely 0:0 contents, or do all unused vectors point to an `IRET`?) for its copy.

In addition, ELKS (and lots of other OSes) trap the timer (`int 8`) interrupt but then will call the old vector to keep some BIOS's happy, so that should be initialized too.

Thank you!

ghaerr commented on blink/blinkenlights.c:
```diff
@@ -2970,9 +2976,11 @@ static void OnVidyaService(void) {
       break;
     case 0x0E:
       OnVidyaServiceTeletypeOutput();
+      Redraw(true);
```
> It seems really inefficient to redraw the whole screen whenever one outputs a single character through int 0x10 function 0xe?

Well, yes, but the entire screen may also be refreshed after each instruction executed, although in C)ontinue mode an effective `Redraw(false)` allows some buffers to be dumped. The inefficiency was why D)isplay mode was created, and as mentioned, it still has some problems. I'm thinking that we could replace the above `Redraw(true)` with `DrawDisplayOnly()` - I'll try that.

> ideally all output will be reflected in the video buffer at 0xb000:0 or 0xb800:0, but currently Blink's pty buffer is separate from the "real mode" memory and the two are not really synchronized.

The pty routines double for program output when not in bare metal mode, it seems the entire pty routine could be replaced with a `DrawEgaText` very similar to `DrawCga`/`DrawMda` and not call `DrawTerminal` at all when not in metal mode. I'll write a routine for that and try it.

ghaerr replied:
Hello @tkchia,

> it seems the entire pty routine could be replaced with a DrawEgaText very similar to DrawCga/DrawMda

I just coded this, and if the `ptyWrite` routine is replaced with a function that writes the char/attr to segment 0xB800 with some control character processing, the entire pty routine can be eliminated for BIOS text modes and the `DrawCga` routine works without modification. The entire emulation should work great if we just set the video mode to 3 when `blink` is started in real mode. 

Here's a screenshot showing what it looks like booting ELKS using this method:
![ELKS CGA](https://user-images.githubusercontent.com/11985637/234959260-27de7004-c3cb-434f-a332-1c4066c47eda.png)

I'll continue working on this and submit a separate PR with 0xB800 EGA and 0xB8000 MDA support worked into the `int 10` AH=9/0Eh function. There may be a few gotcha's if moving from real to long mode, so will implement and test for `blink -r` for the time being.

Thank you!

ghaerr commented on blink/blinkenlights.c:
```diff
@@ -3899,12 +3918,14 @@ static void Tui(void) {
             action &= ~NEXT;
             if (IsCall()) {
               BreakAtNextInstruction();
+              tuimode = false;
```
> what is the particular issue that this change addresses? 

IIRC it allows the control flow to exit the `} while (tuimode);` following the `break;` and exit `Tui()`, then re-enter again through `VirtualMachine`. This allows the display to be shown properly at the breakpoint.

tkchia replied:
@ghaerr : oh OK.  Thanks!

tkchia commented:
Hello @ghaerr,

> Yes, that is definitely needed. Currently ELKS is copying garbage from the uninitialized (likely 0:0 contents, or do all unused vectors point to an `IRET`?) for its copy.

The CP437 dumps should give a rough idea of what is going on.  At the moment, upon "boot up", Blinkenlights arranges for
* interrupts `0`&mdash;`0x1c`
* and interrupts `0x70`&mdash;`0x77` (IRQs 8&mdash;15 on a real PC)

to be vectored to `hvtailcall` instructions.  I think Blinkenlights does not invoke `int 8` (IRQ 0) on its own accord yet.  But if some code suddenly decides to invoke the default `int 8` routine, we will know about it. :smile: 

Thank you!

ghaerr commented:
Hello @tkchia, 

> But if some code suddenly decides to invoke the default int 8 routine, we will know about it. 

Yep, neither `blink` nor `blink16` implement a system timer - yet. That's why the ELKS versions they both now emulate must be compiled with a special CONFIG_TIMER_INT0F feature. This feature replaces the `hlt` instruction in the idle task's main loop with an `INT 0x0f`. In addition, with this option enabled, the kernel hooks IRQ 7 rather than IRQ 0 for the timer interrupt, which then allows the idle loop's `INT 0x0f` to trigger the hardware timer interrupt - without a timer. The clocks are all wrong, but the headless console works by polling the keyboard using `INT 0x16` through a kernel timer, so we get a working shell!

Thank you!

ghaerr commented:
Hello @tkchia,

> At the moment, upon "boot up", Blinkenlights arranges for ... 
> interrupts 0x70—0x77 (IRQs 8—15 on a real PC)
> to be vectored to hvtailcall instructions.

I see. When ELKS gets an actual IRQ 0 interrupt (or simulated `int 8`), the kernel timer update routine is called, which normally would have previously hooked the `int 8` vector. However, after every 5 timer ticks, the original `int 8` vector will be called, which would cause an execution halt if 0:0. (BTW, do we check for 0:0 in `blink`, or just continue and crash? `blink16` checks and halts in that case). I suppose for now we don't want to initialize any unused vectors so we can identify programs that use them. Are you thinking that each would ultimately be replaced with a BIOS `iret`, or an `hvcall`/`hvtailcall` that does nothing?

Thank you!

jart reviewed (changes requested):
I tested this with SectorLISP:

```
make -j16 o//blink && o//blink/blinkenlights -r third_party/sectorlisp/sectorlisp-friendly.bin
```

What happens is that when I press `C` it ends up going wild inserting spaces into the teletypewriter and redraws the display too much. Could you take a look?

ghaerr commented:
> What happens is that when I press C it ends up going wild inserting spaces into the teletypewriter and redraws the display too much. Could you take a look?

Yes, I'll take a look. Thanks for testing. I'm guessing its the `Redraw(true)` I inserted after the `int 10` teletypewriter output function, which @tkchia already mentioned. Do you happen to recall whether SectorLISP uses `int 10` for output? I'm almost sure it would need to.

I'll fix this using a different method, thanks.

jart commented:
SectorLISP uses 16h for keyboard input.

ghaerr commented:
Hello @jart, hello @tkchia,

I rewrote the `DrawDisplayOnly` routine and this should allow for much faster display running SectorLISP with no blinking. Please give it a try.

I also went ahead and implemented BIOS video ram support for all 80x25 text modes (CGA, EGA, MDA). This, along with adding some more `int 10h` functions now allows FreeDOS to fully boot and display perfectly. The way this works is that when `blink` starts in metal mode, it defaults to using the BIOS video ram instead of the PTY routines, which allows intermixed `int 10h` calls and directly writing to video ram. If this is not desired, pressing 'v' at startup will switch back (between) BIOS and PTY drivers.

I've tested this pretty heavily with ELKS, FreeDOS and SectorLISP and all boot and run well. When running in C)ontinue mode, I sometimes see that ^C exits `blink` rather than exiting C)ontinue mode; I'm not sure yet why, I think its to do with executing an `int 16h` read without a poll request first, and that behavior may have already existed as a bug. I will fix that after this PR.

Currently, the BIOS video routines use hard-coded values of 80x25, instead of referencing the BIOS Data Area, which I plan to add in the future. In addition, some quick 6845 CRT controller emulation would allow for full cursor control without using BIOS, which ELKS uses when compiled with Direct Console. There's a few more `int 10h` functions which FreeDOS uses which I plan to add.

Here's a screenshot of the FreeDOS v1.3 installation image at boot:
![FreeDOS EGA](https://user-images.githubusercontent.com/11985637/235054388-a0db4c0b-bae9-4e8d-8b54-8f2941791675.png)

Thank you!

ghaerr commented on blink/blinkenlights.c:
```diff
@@ -1379,14 +1382,21 @@ static void DrawTerminal(struct Panel *p) {
-      DrawCga(p, (u8(*)[80][2])(m->system->real + 0xb8000));
+      DrawCga(p, (u8(*)[80][2])(m->system->real + 0xb8000), pty->x, pty->y);
+      break;
+    case 3:     // EGA 80x25 16-color
```
Thanks, will rename to CGA.

tkchia replied:
Thanks!

ghaerr commented on blink/cga.c:
```diff
@@ -32,17 +33,28 @@ size_t FormatCga(u8 bgfg, char buf[11]) {
-        AppendData(&p->lines[y], buf, FormatCga((a = v[y][x][1]), buf));
+      if (x == curx && y == cury) {
+        AppendData(&p->lines[y], buf, FormatCga(0x07, buf));
+        AppendWide(&p->lines[y], CURSOR);
```
Thank you for your code review :)

I don't think simultaneously using routines from 'blink/cga.c' and `blink/pty.c` will work well for a number of reasons. The pty buffers don't contain character/attribute data including for the cursor, and designing them to be required to be used together adds more complexity than is needed.

I am aware of the problem that the cursor can only be displayed as occupying a space position overwriting any other character "under" it - `blink/pty.c` has the same problem, and uses the same mechanism to display a cursor (which is just the underline (`_`) character in most cases). I am not aware of any technique that allows TUI output to display two glyphs in the same location, do you know how that could be done?

tkchia replied:
> I don't think simultaneously using routines from 'blink/cga.c' and blink/pty.c will work well for a number of reasons.

@ghaerr : but my understanding is that, at the end of the day, the escape sequences from `blink/cga.c` will still go to `blink/pty.c` anyway?

ghaerr replied:
>  but my understanding is that, at the end of the day, the escape sequences from blink/cga.c will still go to blink/pty.c anyway?

No - they operate completely independently. However, both load buffers that are rendered by the code in `blink/panel.c`.

ghaerr replied:
@tkchia: To explain this a bit further - any TUI routines, including `blink/cga.`, `blink/pty.c` and some code in `blink/blinkenlights.c` can include ANSI escape sequences; except for `blink/pty.c`, these sequences are passed through unmodified to `blink/panel.c`, where they are essentially clipped to the panel window and, if visible, sent out unmodified. However, `blink/pty.c` does the additional job of interpreting ANSI escape sequences to simulate an ANSI terminal and then stores the data in a way such that its buffers can be also rendered by `blink/panel.c` by sending final ANSI escape sequences to the real terminal (emulator).

tkchia replied:
@ghaerr : I stand corrected.  In that case I guess what I would propose, is that `blink/cga.c` do something similar to `blink/pty.c`.  That is, if the cursor is on a character cell that is not a space, then use reverse video to display the actual character, rather than using a `▂` glyph.

Thank you!

ghaerr replied:
@tkchia: Good idea, let me see what I can do :)

tkchia replied:
@ghaerr : by the way you need to include either `<termios.h>` or `"blink/xlat.h"` in order to get `IUTF8` defined.  Thank you!

ghaerr replied:
@tkchia: thanks, I was unaware of that. I'll include `<termios.h>`.

ghaerr replied:
@tkchia: The cursor is fixed and tested (using `vframe` FreeDOS command). Thank you!

tkchia commented:
Hello @ghaerr,

> The way this works is that when blink starts in metal mode, it defaults to using the BIOS video ram instead of the PTY routines, which allows intermixed int 10h calls and directly writing to video ram. If this is not desired, pressing 'v' at startup will switch back (between) BIOS and PTY drivers.

Interesting.  Anyway, this seems to be a reinterpretation of what the `v` command was supposed to do (OK &mdash; I must admit I am still not entirely sure of the original intent).

> I suppose for now we don't want to initialize any unused vectors so we can identify programs that use them. Are you thinking that each would ultimately be replaced with a BIOS `iret`, or an `hvcall`/`hvtailcall` that does nothing?

At the moment I mostly follow what the [classical IBM PC AT BIOS](https://archive.org/details/IBMPCATIBM5170TechnicalReference6280070SEP85/page/n237/mode/2up) did at early boot.

The original AT BIOS initialized the vectors for interrupts
* `2` (NMI exception)
* `5` ("print screen" service),
* `8`&mdash;`0x1f` (IRQs 0&mdash;7, software interrupts, and data structures)
* `0x70`&mdash;`0x77` (IRQs 8&mdash;15).

My current `blink/defbios.c` implementation currently initializes the vectors for interrupts `0`&mdash;`0x1c` and `0x70`&mdash;`0x77`.

Thank you!

ghaerr commented:
Hello @tkchia,

> Anyway, this seems to be a reinterpretation of what the v command was supposed to do (OK — I must admit I am still not entirely sure of the original intent).

Me neither - but given the original code was `vidya = !vidya`, I am pretty certain the idea was to "undo" a BIOS video mode switch back to 0 (which used to use the `default:` case) and return to `blink/pty.c` processing. Since BIOS mode 0 is valid, this PR changes the PTY mode to 255 which is otherwise unused (and allows for future BIOS mode 0 which is CGA text 40x25), and using 'v' flips between PTY and BIOS mode 3, which seemed the most useful given the original code.

Thank you!

tkchia commented:
@jart : The changes now look good (well, better :slightly_smiling_face: ) to me.  Do also let me know your opinion.  Thank you!

ghaerr commented:
After testing further with FreeDOS and SectorLISP in both CGA and MDA modes, a few tweaks were made to the new BIOS video functions and CGA/MDA display routines to get the results to almost exactly match QEMU. There are still a few display anomalies, but we're getting pretty close.

I'm going to commit this now before delving deeper into achieving display perfection with QEMU. It is possible that FreeDOS is using the BIOS Data Area contents to check certain things (e.g. display size, etc), and that has not been implemented yet.

BIOS video mode 7 (MDA) is nice since it displays black on white (at least on my macOS terminal, which defaults to black on white for `blink`), which is sometimes interesting. I'm thinking we should add a `-7`  option to `blink` to set mode 7 MDA on metal mode startup instead of the default mode 3 (CGA). Thoughts?
